src/s2client-fsharp/ProtobufConnection.fs

```diff
@@ -1,256 +1,168 @@
-                while true do
-                    let! request = mailbox.Receive()
-                    match request with
-                    | SendRequest (req, reply) -> reply.Reply(ClientResponse.Error catch)
-                    | Disconnect (_, reply) -> reply.Reply(None)                
-            })
-
-        let con = { Client = mailbox; _Address = address; _Port = port; _Timeout = timeout }
-        let! _ = ping con
-        return con }
-
-    let inline checkError (error:'T when 'T : enum<int>) (errorDetails:string) =
-        if error :> obj :?> int <> 0 then
-            raise <| TypedResponseErrorException<'T>(error, errorDetails)
-
-    let createGame (createGame: RequestCreateGame) (cl : Sc2Connection) : Async<Status> = async {
-        let request = new SC2APIProtocol.Request()
-        request.CreateGame <- createGame
-        let! response = sendRequest cl request
-        let createGameResponse = response.CreateGame
-        checkNullAndWarnings response createGameResponse
-        checkError createGameResponse.Error createGameResponse.ErrorDetails
-        return response.Status }
-
-    let joinGame (joinGame: RequestJoinGame) (cl : Sc2Connection) : Async<PlayerId * Status> = async {
-        let request = new SC2APIProtocol.Request()
-        request.JoinGame <- joinGame
-        let! response = sendRequest cl request
-        let joinGameResponse = response.JoinGame
-        checkNullAndWarnings response joinGameResponse
-        checkError joinGameResponse.Error joinGameResponse.ErrorDetails
-        return joinGameResponse.PlayerId, response.Status }
-
-    let getGameInfo (cl : Sc2Connection) = async {
-        let request = new SC2APIProtocol.Request()
-        request.GameInfo <- new RequestGameInfo()
-        let! response = sendRequest cl request
-        let gameInfoResponse = response.GameInfo
-        checkNullAndWarnings response gameInfoResponse
-        return gameInfoResponse, response.Status }
-
-    let getObservation disableFog (cl : Sc2Connection) = async {
-        let request = new SC2APIProtocol.Request()
-        request.Observation <- new RequestObservation()
-        request.Observation.DisableFog <- disableFog
-        let! response = sendRequest cl request
-        let observationResponse = response.Observation
-        checkNullAndWarnings response observationResponse
-        return observationResponse, response.Status }
-
-    let doStep stepSize (cl : Sc2Connection) = async {
-        let request = new SC2APIProtocol.Request()
-        request.Step <- new RequestStep()
-        request.Step.Count <- stepSize
-        let! response = sendRequest cl request
-        let stepResponse = response.Step
-        checkNullAndWarnings response stepResponse
-        return response.Status }
+                    clientSock.ConnectAsync(fullAddress, tok) |> Async.AwaitTask |> Async.RunSynchronously
```

Comment:
Now the `Sc2Connection` constructor blocks and is no longer async. Another design might be 

- change the constructor itself to `private` 
- add an async static `Create/Connect` method
- add the socket to the constructor parameters.

Reply:
There is a connect function later on in the file, that is used to create a Sc2Connection

let connect address port timeout tok = async{
        return Sc2Connection(address, port, timeout, tok)
    }

Is this in line with your comment?